Adding endpoints for retrieving data and SQL for a node

shangyian opened the pull request:
### Summary

This PR adds two endpoints:
* `/data/{node_name}` to retrieve data for a node
* `/sql/{node_name}` to retrieve SQL for a node

These endpoints work for sources, transforms, metrics and dimensions. In order to support data/sql generation for source and transform nodes, I removed the check for a `query` on the node and included a stage that builds the AST for a source node. This also required an additional check to make sure that we don't include any fields in the `SELECT` clause that aren't in the `GROUP BY` clause, when the latter exists. 

I'll work on support for cube nodes in a separate PR.

Other changes:
* Bugfix for generated SQL: When users provide a filter, it shouldn't be included in the `SELECT` clause by default. In addition to not making sense, including it in `SELECT` will also generate a query that doesn't run for metrics or any other node with an aggregation (fails in Postgres with `must appear in the GROUP BY clause or be used in an aggregate function`). 
* Removes `query` table as that is now part of djqs.
* Moved the availability state posting from `/data/availability/{node_name}/` to `/data/{node_name}/availability/` to be consistent with the URL order of `/data/{node_name}/...`
* Added docs to `CONTRIBUTING.rst`

These changes require a minor companion change to DJQS: https://github.com/DataJunction/djqs/pull/10

### Test Plan

Added some tests with a mocked query service client. I also ran this with (a slightly modified) DJQS against the DJ Roads tables.

- [X] PR has an associated issue: #342
- [X] `make check` passes
- [X] `make test` shows 100% unit test coverage

### Deployment Plan

<!-- Any special instructions around deployment? -->

agorajek commented on dj/api/data.py:
```diff
@@ -83,3 +87,46 @@ def add_availability(
+        raise DJException(message=f"Can't get data for node type {node.type}!")
+
+    if node.type == NodeType.DIMENSION:
+        if dimensions or filters:
```
Why not for Dimension node data with filters?

shangyian replied:
Good point, in thinking about this more, we can open this up to filters.

shangyian replied:
Done! I've added support for dimensions with filters.

agorajek commented on dj/api/data.py:
```diff
@@ -83,3 +87,46 @@ def add_availability(
+    Gets data for a node
+    """
+    node = get_node_by_name(session, node_name)
+    if node.type not in (NodeType.METRIC, NodeType.CUBE, NodeType.DIMENSION):
```
@shangyian I am trying to come up with a good reason for not allowing users to query data for a source or transform node but I can't. We can always add a default row limit, on any node type. Just trying to understand this restriction better.

shangyian replied:
I guess I just feel like it's a bit unusual for people to try to access sources or transforms through DJ, but I can also see how if someone is in the system and exploring, it might be useful. But in my view the most powerful part of DJ is its ability to generate queries, and for source nodes that's not going to be applicable. I could see being used for transform nodes though.

agorajek replied:
> But in my view the most powerful part of DJ is its ability to generate queries

Fully agree ^, but in this case I think we don't need to tell people that they can't see data for a source or transform node if they just wanna have a peek at `SELECT * FROM any_node LIMIT 10` or something. It just feels like an artificial restriction, even if this is not a common use case.

shangyian replied:
Makes sense! I just added some changes to support querying both source and transform nodes. It turns out the existing `get_query` doesn't support nodes with no queries, so I added some modifications to generate a select star query AST.

agorajek reviewed (commented):
Nice little api. But really wondering why the restriction?

agorajek commented:
Two in-line comments, but please adjust at your conveniance.

shangyian commented:
@agorajek I added support for sources + transforms. I'll leave the cube nodes to another PR... Also added a /sql endpoint, which serves as a parallel to /data but just returns the SQL without running it.

agorajek commented:
Thanks for the extras @shangyian . Adding query for any node sounds great.
